spec/wrap.md

```diff
@@ -0,0 +1,104 @@
+# WRAP Specification
+
+WRAP defines the rules for bi-directional communication between a Host and a Wasm module by defining the processes of invocation and subinvocation. It makes use of Host allocated Wasm shared memory to pass encoded buffers between the Host and the Wasm module.
+
+The Wasm module must define the following imports and exports:
+
+#### Exports
+- `_invoke: (bufferLen: u32) => u32`
+    - sends the length of input buffer to Wasm module and it returns the length and the ptr to the new 8 byte long response buffer which contains length of the result buffer + pointer to the result buffer
```

Comment:
I think this sentence can be rewritten for better understanding. Maybe "The invoke export function receives the length of the invocation arguments buffer as its first and only argument. The invocation arguments buffer resides in an external host-controlled memory space, and must be copied into the receiving wasm module. To do this, the invoke export function's job is to allocate a new buffer of this size, and return a pointer to it, such that the host can copy the invocations arguments into it.

^^^ this might not be accurate, but you get my point, I think the wording of this paragraph can be massaged.

Reply:
Note: it doesn't return the buffer ptr, but instead it uses the __fill_buffer import fn.

Reply:
Reworded this. The full explanation is in the invocation process below.